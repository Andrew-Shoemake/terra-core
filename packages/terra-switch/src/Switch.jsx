import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import * as KeyCode from 'keycode-js';
import styles from './Switch.module.scss';

const cx = classNames.bind(styles);

// Translations Required
const SWITCH_STATE = Object.freeze({
  ON: 'On',
  OFF: 'Off',
});

const propTypes = {
  /**
   * Id of the Switch button
   */
  id: PropTypes.string.isRequired,
  /**
   * Id of the Switch label
   */
  labelId: PropTypes.string.isRequired,
  /**
   * Whether or not the Switch is checked ("ON").
   */
  checked: PropTypes.bool,
  /**
   * Whether or not the Switch is disabled ("ON").
   */
  disabled: PropTypes.bool,
  /**
   * The label of the Switch component.
   */
  label: PropTypes.node.isRequired,
  /**
   * Callback function when switch value changes from ON / OFF.
   * Parameters: 1. event 2. switch value
   */
  onChange: PropTypes.func,
};

const defaultProps = {
  checked: false,
  disabled: false,
  onChange: undefined,
};

const Switch = ({
  id,
  labelId,
  checked,
  disabled,
  onChange,
  label,
  ...customProps
}) => {
  const sliderButtonRef = useRef(null);

  const handleClick = (event) => {
    sliderButtonRef.current.focus();
    if (onChange) {
      onChange(event, checked);
    }
  };

  const handleOnKeyDown = (event) => {
    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
      event.preventDefault();
    }
  };

  const handleKeyUp = (event) => {
    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
      sliderButtonRef.current.focus();
      if (onChange) {
        onChange(event, checked);
      }
    }
  };

  const switchClassNames = cx([
    'switch',
  ]);

  const trayClassNames = cx([
    'tray',
    { selected: checked },
    { unselected: !checked },
    { 'is-disabled': disabled },
  ]);

  const sliderClassNames = cx([
    'slider',
    { 'is-disabled': disabled },
  ]);

  const statusLabelText = checked ? SWITCH_STATE.ON : SWITCH_STATE.OFF;

  return (
    <div className={cx('switch-wrapper')}>
      <label htmlFor={id}>
        <div className={cx('switch-container')}>
          <div className={cx('label-container')}>
            <span id={labelId} className={cx('label-text')}>{label}</span>
            <span className={cx('status-label-text')}>{statusLabelText}</span>
          </div>
          <div className={switchClassNames}>
            <span className={trayClassNames}>
              <button
                {...customProps}
                type="button"
                id={id}
                disabled={disabled}
                aria-checked={checked}
                aria-labelledby={labelId}
                className={sliderClassNames}
                role="switch"
                tabIndex="0"
                onClick={handleClick}
                onKeyDown={handleOnKeyDown}
                onKeyUp={handleKeyUp}
                ref={sliderButtonRef}
              />
            </span>
          </div>
        </div>
      </label>
    </div>
  );
};

Switch.propTypes = propTypes;
Switch.defaultProps = defaultProps;

export default Switch;
