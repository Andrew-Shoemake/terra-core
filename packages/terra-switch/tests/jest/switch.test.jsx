import React from 'react';
import Switch from '../../src/Switch';

describe('Switch', () => {
  const defaultRender = <Switch labelText="default" />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as disabled when set', () => {
    const button = shallow(<Switch labelText="Label" isDisabled />);
    expect(button).toMatchSnapshot();
  });

  // Prop Tests
  it('should use the default label text when is given', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.find('.label-text').text()).toEqual('default');
  });

  // Structure Tests
  it('should have the class switch', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper.prop('className')).toContain('switch');
  });
});
