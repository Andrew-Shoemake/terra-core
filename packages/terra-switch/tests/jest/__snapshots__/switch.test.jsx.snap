// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render a default component 1`] = `
<div
  className="switch-wrapper"
>
  <label
    htmlFor="terra-switch-button-1"
  >
    <div
      className="switch-container"
    >
      <div
        className="label-container"
      >
        <span
          className="label-text"
          id="switchLabel"
        >
          default
        </span>
        <span
          className="status-label-text"
        >
          Off
        </span>
      </div>
      <div
        className="switch"
      >
        <span
          className="tray unselected"
        >
          <button
            aria-checked={false}
            aria-labelledby="switchLabel"
            className="slider"
            disabled={false}
            id="terra-switch-button-1"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            role="switch"
            tabIndex="0"
            type="button"
          />
        </span>
      </div>
    </div>
  </label>
</div>
`;

exports[`Switch should render as disabled when set 1`] = `
<div
  className="switch-wrapper"
>
  <label
    htmlFor="terra-switch-button-2"
  >
    <div
      className="switch-container"
    >
      <div
        className="label-container"
      >
        <span
          className="label-text"
          id="switchLabel"
        >
          Label
        </span>
        <span
          className="status-label-text"
        >
          Off
        </span>
      </div>
      <div
        className="switch"
      >
        <span
          className="tray unselected is-disabled"
        >
          <button
            aria-checked={false}
            aria-labelledby="switchLabel"
            className="slider is-disabled"
            disabled={true}
            id="terra-switch-button-2"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            role="switch"
            tabIndex="0"
            type="button"
          />
        </span>
      </div>
    </div>
  </label>
</div>
`;
